README.pod
=pod

=head1 NAME

Class::Tiny::Immutable - Minimalist class construction, with read-only
attributes

=head1 SYNOPSIS

In I<Person.pm>:

  package Person;
  
  use Class::Tiny::Immutable qw( name );
  
  1;

In I<Employee.pm>:

  package Employee;
  use parent 'Person';
  
  use Class::Tiny::Immutable qw( ssn ), {
    timestamp => sub { time }   # lazy attribute with default
  };
  
  1;

In I<example.pl>:

  use Employee;
  
  my $obj = Employee->new; # dies, name and ssn attributes are required
  my $obj = Employee->new( name => "Larry", ssn => "111-22-3333" );
  
  my $name = $obj->name;
  my $timestamp = $obj->timestamp;
  
  # no attributes can be set
  $obj->ssn("222-33-4444"); # dies
  $obj->timestamp(time); # dies

=head1 DESCRIPTION

L<Class::Tiny::Immutable> is a wrapper around L<Class::Tiny> which makes the
generated attributes read-only, and required to be set in the object
constructor if they do not have a lazy default defined. In other words,
attributes are either "lazy" or "required".

=head1 METHODS

In addition to methods inherited from L<Class::Tiny>, Class::Tiny::Immutable
defines the following additional introspection method:

=head2 get_all_required_attributes_for

  my @required = Class::Tiny::Immutable->get_all_required_attributes_for($class);

Returns an unsorted list of required attributes known to Class::Tiny::Immutable
for a class and its superclasses.

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2017 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Moo>, L<MooseX::AttributeShortcuts>

=cut


Changes
{{$NEXT}}

0.001     2019-02-10 14:57:55 EST
  - First release


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.009.
use strict;
use warnings;

use 5.006;

use ExtUtils::MakeMaker;

my %WriteMakefileArgs = (
  "ABSTRACT" => "Minimalist class construction, with read-only attributes",
  "AUTHOR" => "Dan Book <dbook\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Class-Tiny-Immutable",
  "LICENSE" => "artistic_2",
  "MIN_PERL_VERSION" => "5.006",
  "NAME" => "Class::Tiny::Immutable",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Class::Tiny" => "1.003"
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "Test::More" => "0.88"
  },
  "VERSION" => "0.001",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Class::Tiny" => "1.003",
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Test::More" => "0.88"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);


lib/Class/Tiny/Immutable.pm
use strict;
use warnings;

package Class::Tiny::Immutable;

use Carp ();
use Class::Tiny ();
our @ISA = 'Class::Tiny';

our $VERSION = '0.002';

my %REQUIRED_ATTRIBUTES;

sub prepare_class {
  my ( $class, $pkg ) = @_;
  no strict 'refs';
  @{"${pkg}::ISA"} = "Class::Tiny::Immutable::Object" unless @{"${pkg}::ISA"};
}

sub create_attributes {
  my ( $class, $pkg, @spec ) = @_;
  $class->SUPER::create_attributes( $pkg, @spec );
  $REQUIRED_ATTRIBUTES{$pkg}{$_} = 1 for grep { ref $_ ne 'HASH' } @spec;
}

sub __gen_sub_body {
  my ($self, $name, $has_default, $default_type) = @_;
  
  if ($has_default && $default_type eq 'CODE') {
    return <<"HERE";
sub $name {
  return (
      ( \@_ == 1 )
    ? ( exists \$_[0]{$name} ? \$_[0]{$name} : ( \$_[0]{$name} = \$default->( \$_[0] ) ) )
    : Carp::croak( "$name is a read-only accessor" )
  );
}
HERE
  }
  elsif ($has_default) {
    return <<"HERE";
sub $name {
  return (
      ( \@_ == 1 )
    ? ( exists \$_[0]{$name} ? \$_[0]{$name} : ( \$_[0]{$name} = \$default ) )
    : Carp::croak( "$name is a read-only accessor" )
  );
}
HERE
  }
  else {
    return <<"HERE";
sub $name {
  return \@_ == 1 ? \$_[0]{$name} : Carp::croak( "$name is a read-only accessor" );
}
HERE
  }
}

sub get_all_required_attributes_for {
  my ( $class, $pkg ) = @_;
  # attributes are stored per package, so we need to walk the mro ourselves
  # rely on Class::Tiny to have loaded the appropriate mro
  my %attr =
    map { $_ => undef }
    map { keys %{ $REQUIRED_ATTRIBUTES{$_} || {} } } @{ mro::get_linear_isa($pkg) };
  return keys %attr;
}

package Class::Tiny::Immutable::Object;

our @ISA = 'Class::Tiny::Object';

our $VERSION = '0.002';

sub BUILD {
  my ( $self, $args ) = @_;
  my @missing = grep { !exists $args->{$_} }
    Class::Tiny::Immutable->get_all_required_attributes_for( ref $self );
  Carp::croak( 'Missing required attributes: ' . join( ', ', sort @missing ) ) if @missing;
}

1;

=head1 NAME

Class::Tiny::Immutable - Minimalist class construction, with read-only
attributes

=head1 SYNOPSIS

In I<Person.pm>:

  package Person;
  
  use Class::Tiny::Immutable qw( name );
  
  1;

In I<Employee.pm>:

  package Employee;
  use parent 'Person';
  
  use Class::Tiny::Immutable qw( ssn ), {
    timestamp => sub { time }   # lazy attribute with default
  };
  
  1;

In I<example.pl>:

  use Employee;
  
  my $obj = Employee->new; # dies, name and ssn attributes are required
  my $obj = Employee->new( name => "Larry", ssn => "111-22-3333" );
  
  my $name = $obj->name;
  my $timestamp = $obj->timestamp;
  
  # no attributes can be set
  $obj->ssn("222-33-4444"); # dies
  $obj->timestamp(time); # dies

=head1 DESCRIPTION

L<Class::Tiny::Immutable> is a wrapper around L<Class::Tiny> which makes the
generated attributes read-only, and required to be set in the object
constructor if they do not have a lazy default defined. In other words,
attributes are either "lazy" or "required".

=head1 METHODS

In addition to methods inherited from L<Class::Tiny>, Class::Tiny::Immutable
defines the following additional introspection method:

=head2 get_all_required_attributes_for

  my @required = Class::Tiny::Immutable->get_all_required_attributes_for($class);

Returns an unsorted list of required attributes known to Class::Tiny::Immutable
for a class and its superclasses.

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2017 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Moo>, L<MooseX::AttributeShortcuts>


lib/Class/Tiny/Lazy.pm
use strict;
use warnings;

package Class::Tiny::Lazy;

use Carp ();
use Class::Tiny ();
our @ISA = 'Class::Tiny';

our $VERSION = '0.001';

sub prepare_class {
  my ( $class, $pkg ) = @_;
  no strict 'refs';
  @{"${pkg}::ISA"} = "Class::Tiny::Lazy::Object" unless @{"${pkg}::ISA"};
}

sub __gen_sub_body {
  my ($self, $name, $has_default, $default_type) = @_;
  
  if ($has_default && $default_type eq 'CODE') {
    return <<"HERE";
sub $name {
  return (
      ( \@_ == 1 )
    ? ( exists \$_[0]{$name} ? \$_[0]{$name} : ( \$_[0]{$name} = \$default->( \$_[0] ) ) )
    : Carp::croak( "$name is a read-only accessor" )
  );
}
HERE
  }
  elsif ($has_default) {
    return <<"HERE";
sub $name {
  return (
      ( \@_ == 1 )
    ? ( exists \$_[0]{$name} ? \$_[0]{$name} : ( \$_[0]{$name} = \$default ) )
    : Carp::croak( "$name is a read-only accessor" )
  );
}
HERE
  }
  else {
    return <<"HERE";
sub $name {
  return \@_ == 1 ? \$_[0]{$name} : Carp::croak( "$name is a read-only accessor" );
}
HERE
  }
}

package Class::Tiny::Lazy::Object;

our @ISA = 'Class::Tiny::Object';

our $VERSION = '0.001';

sub BUILD {
  my ( $self, $args ) = @_;
  my $attrs = Class::Tiny::Lazy->get_all_attribute_defaults_for( ref $self );
  my @missing;
  foreach my $name ( keys %$attrs ) {
    # Any attribute without a defined default is required
    push @missing, $name unless defined $attrs->{$name} or exists $args->{$name};
  }
  Carp::croak( 'Missing required attributes: ' . join( ', ', sort @missing ) ) if @missing;
}

1;

=head1 NAME

Class::Tiny::Lazy - Minimalist class construction, with lazy attributes

=head1 SYNOPSIS

In I<Person.pm>:

  package Person;
  
  use Class::Tiny::Lazy qw( name );
  
  1;

In I<Employee.pm>:

  package Employee;
  use parent 'Person';
  
  use Class::Tiny::Lazy qw( ssn ), {
    timestamp => sub { time }   # lazy attribute with default
  };
  
  1;

In I<example.pl>:

  use Employee;
  
  my $obj = Employee->new; # dies, name and ssn attributes are required
  my $obj = Employee->new( name => "Larry", ssn => "111-22-3333" );
  
  my $name = $obj->name;
  my $timestamp = $obj->timestamp;
  
  # no attributes can be set
  $obj->ssn("222-33-4444"); # dies
  $obj->timestamp(time); # dies

=head1 DESCRIPTION

L<Class::Tiny::Lazy> is a wrapper around L<Class::Tiny> which makes the
generated attributes read-only, and required to be set in the object
constructor if they do not have a lazy default defined. In other words,
attributes are either "lazy" or "required".

=head1 BUGS

Report any issues on the public bugtracker.

=head1 AUTHOR

Dan Book <dbook@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2017 by Dan Book.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=head1 SEE ALSO

L<Moo>, L<MooseX::AttributeShortcuts>
